## Text Analysis - Companies description 

In this notebook, we will do some text analysis for the companies existing in the WRDS and Harvard Business School, Impact-Weighted Accounts Project report.

We will apply some Nature Language Processing (NLP) using the pre-trained DistilBERT. First, we need to merge the datasets and obtain the companies description from Yahoo Finance. 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import missingno as msno
import warnings
from sklearn import linear_model
from sklearn import metrics
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import requests
from bs4 import BeautifulSoup
warnings.filterwarnings('ignore')

In [2]:
df_ishares = pd.read_csv('/Users/maralinetorres/Documents/GitHub/Predicting-Environmental-and-Social-Actions/Datasets/iShares-merged.csv')
df_ei = pd.read_csv('/Users/maralinetorres/Documents/GitHub/Predicting-Environmental-and-Social-Actions/Datasets/Environmental_impact_cleaned.csv')
stocks = pd.read_csv("/Users/maralinetorres/Documents/GitHub/Predicting-Environmental-and-Social-Actions/Datasets/pilot_stocks.csv")

For the ishares, we don't need the tickers in the sector 'Cash and devirates' and we are also going to filter by the necessary columns

In [3]:
df_ishares = df_ishares.loc[df_ishares.Sector != 'Cash and/or Derivatives',['Ticker','Name','Sector','CUSIP','ISIN','Location']]
df_ishares.drop_duplicates(inplace=True)
df_ishares.head()

Ticker                  Name                  Sector      CUSIP  \
0   AAPL             APPLE INC  Information Technology   37833100   
1   MSFT        MICROSOFT CORP  Information Technology  594918104   
2   AMZN        AMAZON COM INC  Consumer Discretionary   23135106   
3   TSLA             TESLA INC  Consumer Discretionary  88160R101   
4     FB  FACEBOOK CLASS A INC           Communication  30303M102   

           ISIN       Location  
0  US0378331005  United States  
1  US5949181045  United States  
2  US0231351067  United States  
3  US88160R1014  United States  
4  US30303M1027  United States

In [4]:
len(df_ishares.Ticker.unique())

855

In [5]:
df_ei2 = df_ei.copy()
df_ei2 = df_ei2.loc[:,['ISIN','CompanyName','Country']]
df_ei2.drop_duplicates(inplace=True)
df_ei2.head()

ISIN         CompanyName         Country
0  GB00BMX64W89            Saga plc  United Kingdom
1  MYL1818OO003  BURSA MALAYSIA BHD        Malaysia
2  GB0031638363  INTERTEK GROUP PLC  United Kingdom
3  ZAE000079711         JSE LIMITED    South Africa
4  FR0006174348   BUREAU VERITAS SA          France

In [6]:
len(df_ei2.ISIN.unique())

2628

In [7]:
stocks = stocks.iloc[:,0:2]
stocks.head()

Year Ticker
0  2005    AEE
1  2006    AEE
2  2007    AEE
3  2008    AEE
4  2009    AEE

Let's start merging the datasets. First, we are going to merge iShares and Environmental impact by ISIN. 

In [8]:
df = pd.merge(df_ishares, df_ei2, on='ISIN', how='outer')
df.head()

Ticker                  Name                  Sector      CUSIP  \
0   AAPL             APPLE INC  Information Technology   37833100   
1   MSFT        MICROSOFT CORP  Information Technology  594918104   
2   AMZN        AMAZON COM INC  Consumer Discretionary   23135106   
3   TSLA             TESLA INC  Consumer Discretionary  88160R101   
4     FB  FACEBOOK CLASS A INC           Communication  30303M102   

           ISIN       Location             CompanyName        Country  
0  US0378331005  United States                     NaN            NaN  
1  US5949181045  United States   MICROSOFT CORPORATION  United States  
2  US0231351067  United States        AMAZON.COM, INC.  United States  
3  US88160R1014  United States                     NaN            NaN  
4  US30303M1027  United States  FACEBOOK INCORPORATION  United States

In [9]:
print(df.shape)
print(len(df.Ticker.unique()))

(3005, 8)
856


In [10]:
mising_Ticker = df.loc[df.Ticker.isna(),'ISIN']
df_missing = df_ei.loc[df_ei.ISIN.isin(mising_Ticker), ['ISIN','CompanyName','Country']]
df_missing.drop_duplicates(inplace=True)
print(f'We are unable to match {df_missing.shape[0]} ISIN')
df_missing.head()

We are unable to match 2136 ISIN


ISIN         CompanyName         Country
0  GB00BMX64W89            Saga plc  United Kingdom
1  MYL1818OO003  BURSA MALAYSIA BHD        Malaysia
3  ZAE000079711         JSE LIMITED    South Africa
5  GB0007370074         RICARDO PLC  United Kingdom
6  AU000000ASX7         ASX LIMITED       Australia

With this merge, we were able to find a 492 tickers for the 2,628 companies in the Environmental Impact dataset. We need to figure out how to map their ISIN to a Ticker.

In [11]:
df_missing.to_csv('Companies_missing_Tickers.csv', index=False)

In the meantime, we are going to export the companies that we were unable to match and continue working with the ones that have the Ticker.

From the pilot stocks, we have 52 unique tickers. Let's see if all these tickers appear in the merged dataframe. 

In [12]:
tickers = stocks.Ticker.unique()
x = df.loc[(df.Ticker.isin(tickers)), ['Ticker']]
print(x.shape)
print(f'{len(x.Ticker.unique())} Tickers')
x[x.duplicated()]

(55, 1)
52 Tickers


Ticker
540    DTE
726    MRO
802    ATO

We noticed that these three tickers are duplicated. Let's see why?

In [13]:
df_ishares[df_ishares.Ticker == 'ATO']

Ticker               Name                  Sector      CUSIP  \
417    ATO  ATMOS ENERGY CORP               Utilities   49560105   
809    ATO               ATOS  Information Technology  S56547813   

             ISIN       Location  
417  US0495601058  United States  
809  FR0000051732         France

It seems the ticker is not globally unique. We are going to filter by United States because our pilot stocks belongs to the SP500 US Stocks. 

In [14]:
df_us = df.loc[(df.Ticker.isin(tickers)) & (df.Location == 'United States'), ['Ticker']]
print(f'Now, we have {len(df_us.Ticker.unique())} tickers.')

Now, we have 52 tickers.


We know that we weren't able to match the ISIN for all the companies in the Environmental Intensity dataset. However, let's see the ones that we were able too and validate they matched correctly. 

In [15]:
found = df.loc[(df.Ticker.notna()) & (df.CompanyName.notna()),]
print(len(found.Ticker.unique()))
print(found.shape)

487
(495, 8)


Could it be possible to have duplicate values? Let's subset for the records that have the same ticker.

In [16]:
ticker_agg = found.groupby('Ticker')[['Name']].count().sort_values(by='Name', ascending=False).reset_index()
tickers2 = ticker_agg[ticker_agg.Name >= 2]['Ticker']
df3 = found[found.Ticker.isin(tickers2)].sort_values(by='Ticker')
df3.head(6)

Ticker                           Name                  Sector      CUSIP  \
428    AAL    AMERICAN AIRLINES GROUP INC             Industrials  02376R102   
557    AAL             ANGLO AMERICAN PLC               Materials          -   
99     ADP  AUTOMATIC DATA PROCESSING INC  Information Technology   53015103   
866    ADP          AEROPORTS DE PARIS SA             Industrials          -   
271    DTE                     DTE ENERGY               Utilities  233331107   
540    DTE          DEUTSCHE TELEKOM N AG           Communication  S58423591   

             ISIN        Location                      CompanyName  \
428  US02376R1023   United States      AMERICAN AIRLINES GROUP INC   
557  GB00B1XZS820  United Kingdom               ANGLO AMERICAN PLC   
99   US0530151036   United States  AUTOMATIC DATA PROCESSING, INC.   
866  FR0010340141          France               AEROPORTS DE PARIS   
271  US2333311072   United States               DTE ENERGY COMPANY   
540  DE0005557508         Germany              DEUTSCHE TELEKOM AG   

            Country  
428   United States  
557  United Kingdom  
99    United States  
866          France  
271   United States  
540         Germany

It happened the same thing as before, it seems the Ticker is not globally unique. However, it seems it matches correctly by ISIN. We did a little research and found that AAL in UK actually is AAL.L in Yahoo Finance. 

Another example,  DTE- DEUTSCHE TELEKOM ticker in Yahoo Finance is DTE.DE.  We decided to re-format the Ticker for the companies that are not from US. To some of them, we need to add '.' + 'the first two letters from the ISIN' to the current ticker. 

Let's see. 

In [17]:
ticker_agg = df.groupby('Ticker')[['Name']].count().sort_values(by='Name', ascending=False).reset_index()
tickers2 = ticker_agg[ticker_agg.Name >= 2]['Ticker']
df3 = df[df.Ticker.isin(tickers2)].sort_values(by='Ticker')
df3.head(6)

Ticker                           Name                  Sector      CUSIP  \
428    AAL    AMERICAN AIRLINES GROUP INC             Industrials  02376R102   
557    AAL             ANGLO AMERICAN PLC               Materials          -   
735    ADM              ADMIRAL GROUP PLC              Financials          -   
234    ADM         ARCHER DANIELS MIDLAND        Consumer Staples   39483102   
866    ADP          AEROPORTS DE PARIS SA             Industrials          -   
99     ADP  AUTOMATIC DATA PROCESSING INC  Information Technology   53015103   

             ISIN        Location                      CompanyName  \
428  US02376R1023   United States      AMERICAN AIRLINES GROUP INC   
557  GB00B1XZS820  United Kingdom               ANGLO AMERICAN PLC   
735  GB00B02J6398  United Kingdom                ADMIRAL GROUP PLC   
234  US0394831020   United States                              NaN   
866  FR0010340141          France               AEROPORTS DE PARIS   
99   US0530151036   United States  AUTOMATIC DATA PROCESSING, INC.   

            Country  
428   United States  
557  United Kingdom  
735  United Kingdom  
234             NaN  
866          France  
99    United States

In [18]:
df4 = df3[df3.Location != 'United States']
df4

Ticker                    Name                  Sector      CUSIP  \
557    AAL      ANGLO AMERICAN PLC               Materials          -   
735    ADM       ADMIRAL GROUP PLC              Financials          -   
866    ADP   AEROPORTS DE PARIS SA             Industrials          -   
802    ATO                    ATOS  Information Technology  S56547813   
553     DG                VINCI SA             Industrials          -   
540    DTE   DEUTSCHE TELEKOM N AG           Communication  S58423591   
560     EL     ESSILORLUXOTTICA SA  Consumer Discretionary  S72124779   
731    LEG     LEG IMMOBILIEN N AG             Real Estate          -   
631    MRK                   MERCK             Health Care  S47418447   
726    MRO  MELROSE INDUSTRIES PLC             Industrials          -   
563    PRU          PRUDENTIAL PLC              Financials  S07099542   
547    SAN      BANCO SANTANDER SA              Financials  S57059461   
516    SAN               SANOFI SA             Health Care  S56717358   
739    TEL                 TELENOR           Communication  S47324959   
586   TSCO               TESCO PLC        Consumer Staples  S08847097   

             ISIN        Location             CompanyName         Country  
557  GB00B1XZS820  United Kingdom      ANGLO AMERICAN PLC  United Kingdom  
735  GB00B02J6398  United Kingdom       ADMIRAL GROUP PLC  United Kingdom  
866  FR0010340141          France      AEROPORTS DE PARIS          France  
802  FR0000051732          France                 ATOS SE          France  
553  FR0000125486          France                   VINCI          France  
540  DE0005557508         Germany     DEUTSCHE TELEKOM AG         Germany  
560  FR0000121667          France     ESSILORLUXOTTICA SA          France  
731  DE000LEG1110         Germany       LEG IMMOBILIEN AG         Germany  
631  DE0006599905         Germany              MERCK KGAA         Germany  
726  GB00BZ1G4322  United Kingdom  MELROSE INDUSTRIES PLC  United Kingdom  
563  GB0007099541  United Kingdom          PRUDENTIAL PLC  United Kingdom  
547  ES0113900J37           Spain      BANCO SANTANDER SA           Spain  
516  FR0000120578          France             SANOFI S.A.          France  
739  NO0010063308          Norway       TELENOR GROUP ASA          Norway  
586  GB0008847096  United Kingdom               TESCO PLC  United Kingdom

In [19]:
ticker = df4.Ticker.tolist()
yahoo_ticker = ['AAL.L','ADM.L','ADP.PA','ATO.PA','DG.PA','DTE.DE','EL.PA','LEG.DE','MRK.DE','MRO.L','PRU.L','SAN.PA','SAN.MC','TEL.OL','TSCO.L']

In [20]:
df = df.sort_values(by='Ticker')
df.loc[(df.Ticker.isin(ticker)) & (df.Location != 'United States'), 'Ticker'] = yahoo_ticker
df.head()

Ticker                         Name       Sector      CUSIP          ISIN  \
867      -                 IBERDROLA SA    Utilities          -  ES0144583236   
711   1COV                  COVESTRO AG    Materials          -  DE0006062144   
170      A     AGILENT TECHNOLOGIES INC  Health Care  00846U101  US00846U1016   
428    AAL  AMERICAN AIRLINES GROUP INC  Industrials  02376R102  US02376R1023   
557  AAL.L           ANGLO AMERICAN PLC    Materials          -  GB00B1XZS820   

           Location                  CompanyName         Country  
867           Spain                          NaN             NaN  
711         Germany                  COVESTRO AG         Germany  
170   United States   AGILENT TECHNOLOGIES, INC.   United States  
428   United States  AMERICAN AIRLINES GROUP INC   United States  
557  United Kingdom           ANGLO AMERICAN PLC  United Kingdom

In [21]:
print(len(df.Ticker.unique()))

870


Now, we have 870 unique tickers that we can send to yahoo finance and get the company description. Below, we will create a tickers list.

In [25]:
tickers = df.Ticker.unique().tolist()

### Collect companies descriptions

In this section, we will grab the company description directly from the Yahoo Finance website for the 870 companies

In [46]:
import yfinance as yf


ticker_example = tickers[:5]
for t in ticker_example:
    ticker = yf.Ticker(t)
    info = None
    try:
        info = ticker.info
    except:
        print(f"Cannot get info of {t}, it probably does not exist")
        continue
  
  # Got the info of the ticker, do more stuff with it
  #print(f"Info of {t}: {info}")

Cannot get info of -, it probably does not exist
Cannot get info of 1COV, it probably does not exist
Cannot get info of A, it probably does not exist
Cannot get info of AAL, it probably does not exist
Cannot get info of AAL.L, it probably does not exist


In [53]:
# Create a loop to store URLs of all stocks' description page
URL = [] 
DES = [] 
comp_desc = {}
for i in tickers[:2]: 
    i = str(i)
    url ='https://finance.yahoo.com/quote/'+i+'/profile' 
    URL.append(url)
    page = requests.get(url) # visits the URL 
    htmldata = BeautifulSoup(page.content, 'html.parser')
    Business_Description = htmldata.find('p',{'class':'Mt(15px) Lh(1.6)'}) # finds the business description part in the HTML code
    DES.append(Business_Description.text)
    comp_desc[i] = [Business_Description.text]

AttributeError: 'NoneType' object has no attribute 'text'

In [51]:
df_comp_desc = pd.DataFrame.from_dict(comp_desc, orient='index', columns = ['Description'])
df_comp_desc.reset_index(inplace = True)
df_comp_desc.rename(columns = {'index':'Ticker'}, inplace=True)
df_comp_desc.head()

Ticker Description
0      -        None
1   1COV        None
2      A        None
3    AAL        None
4  AAL.L        None

In [52]:
df_comp_desc.to_csv('Companydescription.csv',index=False)

### Start working with the DistilBERT

If you don't have the package install, you might need to run: !pip install transformers

If you do, only import the packcage and continue working. 